### ISSUE TITLE
"fixFirefoxAnchorBug" function in doctools.js causes navigating problem in Py3 doc in Chrome

### ISSUE THREAD

**Issue by f8230e64-4d73-4251-8e5a-7f32e4674e2d** [python/cpython#75849]
BPO | [31668](https://bugs.python.org/issue31668)
--- | :---
Nosy | @berkerpeksag, @fireattack, @Mariatta

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-11-04.11:17:18.937>
created_at = <Date 2017-10-03.04:35:19.983>
labels = ['3.8', 'type-bug', '3.7', 'docs']
title = '"fixFirefoxAnchorBug" function in doctools.js causes navigating problem in Py3 doc in Chrome'
updated_at = <Date 2017-11-04.11:17:18.935>
user = 'https://github.com/fireattack'
```

bugs.python.org fields:
```python
activity = <Date 2017-11-04.11:17:18.935>
actor = 'berker.peksag'
assignee = 'docs@python'
closed = True
closed_date = <Date 2017-11-04.11:17:18.937>
closer = 'berker.peksag'
components = ['Documentation']
creation = <Date 2017-10-03.04:35:19.983>
creator = 'fireattack'
dependencies = []
files = []
hgrepos = []
issue_num = 31668
keywords = []
message_count = 4.0
messages = ['303576', '303629', '303631', '305555']
nosy_count = 4.0
nosy_names = ['docs@python', 'berker.peksag', 'fireattack', 'Mariatta']
pr_nums = []
priority = 'normal'
resolution = 'third party'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue31668'
versions = ['Python 3.4', 'Python 3.5', 'Python 3.6', 'Python 3.7', 'Python 3.8']
```

</p></details>

**Comment by f8230e64-4d73-4251-8e5a-7f32e4674e2d** [comment]
Problem

This is a regression bug/flaw in Sphinx's doctools.js, a JS file used in its base template, and therefore got inherited to Python 3's documentation website. Python 2's documentation website is not affected because it's based on an older version of Sphinx.

There is a function in doctools.js:

/**
* workaround a firefox stupidity
* see: https://bugzilla.mozilla.org/show_bug.cgi?id=645075
\*/
fixFirefoxAnchorBug : function() {
if (document.location.hash)
window.setTimeout(function() {
document.location.href += '';
}, 10);
}


This function was supposed to fix an anchor bug in Firefox (see comment). It \*used to* have a condition of $.browser outside, so it will only be applied to Firefox; but it was removed in JQuery so it was removed here as well. Therefore, this function now applies to all the browsers. Unfortunately, it causes navigating problem in Chrome now, when you use back and forward.

The problem STR (Chrome only):

1. Open a link with hash (anchor), e.g. https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str
2. Scroll away from the anchor position.
3. Click any external link (means the link that is not an anchor of current page).
4. Hit "back" button in the browser.

What happened:

When you navigating back, it doesn't go to your previous position. Instead, it goes to the anchor's location.

What it should do:

It should go to your previous position.

Ironically, it won't cause this problem in Firefox, despite it's supposed to be a fix for (a different) anchor bug in Firefox.

Comments

I reported it to Sphinx as well: https://github.com/sphinx-doc/sphinx/issues/3549 but didn't get any response so far.

Please keep in mind the Firefox anchor bug mentioned above will only happen if the anchor is assigned with \<section id\>. From my observation we don't really use \<section\> in HTML, so I don't think this workaround function is really necessary here to begin with.

**Comment by Mariatta** [comment]
Is there anything to be done in CPython? It appears to be either a bug with the browser or Sphinx...

**Comment by f8230e64-4d73-4251-8e5a-7f32e4674e2d** [comment]
Well, since it affects Python's site's functionality, I thought it's appropriate to report here (with components: documentation). I followed the instruction here: https://docs.python.org/3/bugs.html#documentation-bugs

It is not Sphinx's bug per se, since it has nothing to do with its software; just an ancient problematic script in template got copied here. 

It is not a Chrome bug, since that script was not intended to apply to Chrome to begin with.

But anyway, if anyone found it's not the right place, feel free to close it.

**Comment by berkerpeksag** [comment]
This is indeed an annoying bug. Turns out there is already a compatibility shim for $.browser in doctools.js: https://github.com/sphinx-doc/sphinx/commit/c608af4babe140626877be08535af095ff633c00

I've opened https://github.com/sphinx-doc/sphinx/pull/4216 to add the "$.browser.firefox" check back.

There is nothing we can do here because we don't maintain a copy of doctools.js in Python codebase:

```sh
    $ git ls-files | grep doctools.js
```


Thank you for your report.
